### PR TITLE
Add additional safeguards for config write access

### PR DESCRIPTION
**Pull request by benhalpern**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This related to this issue: #6049

I wouldn't want to say it fully closes it, but is a step in safeguarding this area.

This pull request adds extra safeguards for the `/internal/config` area of the site. This area contains values which could cause big problems if accidentally changed. It makes sense to restrict them even more thoroughly than other parts of the internal experience.

It now makes it so that anyone with `admin`,  `super_admin` or `single_resource_admin Config` can _view_ this area, but only users with the role `single_resource_admin Config` can write to it. And they need to submit this phrase when making a change as confirmation:

`My username is @your_username and this action is 100% safe and appropriate.`

The details here really should not change very often, so I think this is all pretty reasonable.

Added a message to make it clear:

Screen Shot 2020-02-13 at 1 22 20 PM

Screen Shot 2020-02-13 at 1 21 52 PM

Screen Shot 2020-02-13 at 1 30 45 PM

Also made a small tweak to `/internal/permissions` to make it clearer which exact permissions folks have.

Screen Shot 2020-02-13 at 1 22 51 PM

### CODE REVIEW

File: app/controllers/internal/configs_controller.rb [hunk 1]
```diff
@@ -31,6 +32,11 @@ def config_params
     params.require(:site_config).permit(allowed_params)
   end
 
+  def confirmation
+    not_authorized unless current_user.has_role?(:single_resource_admin, Config) # Special additional permission
```
Comment (benhalpern) [inline app/controllers/internal/configs_controller.rb]: I added this here instead of via pundit because this seemed like an explicit additional step here, and colocated it with the other confirmation check. Open to suggestions if this is not ideal.

File: app/views/layouts/internal.html.erb [hunk 2]
```diff
@@ -40,7 +40,7 @@
 
   <div class="container mt-3">
     <% flash.each do |type, message| %>
-      <div class="alert alert-<%= type %>">
+      <div class="alert alert-<%= type == "notice" ? "success" : "danger" %>">
```
Comment (benhalpern) [inline app/views/layouts/internal.html.erb]: This makes sure we get the right color here. Not really specific to the issue but the old one wasn't quite right.

File: app/controllers/internal/configs_controller.rb (hunk 1 above)
Comment (vaidehijoshi) [reply app/controllers/internal/configs_controller.rb]: any reason to not make a it a `before_action` for the `create` action here and put inline?

File: app/controllers/internal/configs_controller.rb [hunk 3]
```diff
@@ -31,6 +32,11 @@ def config_params
...
+    not_authorized if params[:confirmation] != "My username is @#{current_user.username} and this action is 100% safe and appropriate."
```
Comment (vaidehijoshi) [inline app/controllers/internal/configs_controller.rb]: will `not_authorized` on this line tell the user that their confirmation was incorrect? I'm specifically thinking if someone typos this confirmation; we might want to explicitly tell them that their confirmation was not correct (instead of the whatever `not_authorized` responds with?)

File: app/views/internal/configs/show.html.erb [hunk 4]
```diff
@@ -1,3 +1,16 @@
+<div class="alert alert-warning" role="alert">
+  <p>
+    Only admins with explicit <strong>single resource admin</strong> role may make changes to site config. Even super admins require this permission.
```
Comment (vaidehijoshi) [inline app/views/internal/configs/show.html.erb]:
```suggestion
    Only admins with an explicit <strong>single resource admin</strong> role may make changes to site config. Even super admins require this permission.
```

File: app/views/internal/configs/show.html.erb [hunk 5]
```diff
@@ -1,3 +1,16 @@
...
+  <p>
+  <h2>
+    <% if current_user.has_role?(:single_resource_admin, Config) %>
+      <div class="alert alert-primary">You have permission to make config changes. Do so with care.</div>
```
Comment (vaidehijoshi) [inline app/views/internal/configs/show.html.erb]: 👍

File: spec/requests/internal/configs_spec.rb [hunk 6]
```diff
@@ -15,137 +17,166 @@
...
-    describe "Tags" do
-      it "removes space suggested_tags" do
-        post "/internal/config", params: { site_config: { suggested_tags: "hey, haha,hoho, bobo fofo" } }
-        expect(SiteConfig.suggested_tags).to eq(%w[hey haha hoho bobofofo])
-      end
-
-      it "downcases suggested_tags" do
-        post "/internal/config", params: { site_config: { suggested_tags: "hey, haha,hoHo, Bobo Fofo" } }
-        expect(SiteConfig.suggested_tags).to eq(%w[hey haha hoho bobofofo])
+    context "when admin has full permissions" do
```
Comment (vaidehijoshi) [inline spec/requests/internal/configs_spec.rb]:
```suggestion
    context "when admin has single-resource and full permissions" do
```
This is slightly different than just full permissions, right?

File: spec/requests/internal/configs_spec.rb [hunk 7]
```diff
@@ -15,137 +17,166 @@
...
+      before do
+        sign_in(admin_plus_config)
+      end
+      describe "staff" do
+        it "does not allow the staff_user_id to be updated" do
+          expect(SiteConfig.staff_user_id).to eq(1)
+          post "/internal/config", params: { site_config: { staff_user_id: 2 }, confirmation: confirmation_message }
+          expect(SiteConfig.staff_user_id).to eq(1)
+        end
```
Comment (vaidehijoshi) [inline spec/requests/internal/configs_spec.rb]: do we want to add this back in to the config form? or will that be a separate PR?

File: app/controllers/internal/configs_controller.rb (hunk 1 above)
Comment (benhalpern) [reply app/controllers/internal/configs_controller.rb]: Hmmm. I had put it here to be more explicit, but since we do this with other kinds of authorization that does make sense, I'll make this change.

File: spec/requests/internal/configs_spec.rb (hunk 6 above)
Comment (benhalpern) [reply spec/requests/internal/configs_spec.rb]: Good call. But I'm going to make a different change instead of this exact commit.

File: spec/requests/internal/configs_spec.rb (hunk 7 above)
Comment (benhalpern) [reply spec/requests/internal/configs_spec.rb]: I feel like we can make this a separate thing if we want to. Might be nice to sit on this before making the final choice.

Comment (vaidehijoshi) [reply spec/requests/internal/configs_spec.rb]: that sounds good to me 👍

File: app/controllers/internal/configs_controller.rb (hunk 3 above)
Comment (benhalpern) [reply app/controllers/internal/configs_controller.rb]: This is definitely the ideal behavior.

I felt like because of the fairly custom logic going on here with `SiteConfig.public_send` and how possibly destructive this is, I felt kind of nervous about trying to get the UX _just right_ in this step as opposed to just blowing things up at this step if it's not done correctly.

I think a follow-on PR to get this flow proper with plenty of testing to ensure this absolutely can't go wrong would be a good call.

File: spec/requests/internal/configs_spec.rb (hunk 7 above)
Comment (vaidehijoshi) [reply spec/requests/internal/configs_spec.rb]: @benhalpern I'm looking at this code again, and I'm wondering if we meant for the `staff_user_id` to _not_ be updated, even if you pass it a `confirmation_message`? If someone passes a confirmation message, we should allow for this to be updated, right?

Just pairing with @juliannatetreault on adding `mascot_user_id` into the SiteConfig right now and noticed that `staff_user_id` wasn't in the params, so wanted to double-check the intent here!
